File: lib/services/pro_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cloud_functions/cloud_functions.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'cached_readers.dart';

/// Manages AnySkill Pro status.
///
/// Firestore:
///   system_settings/pro  — thresholds doc
///   users/{uid}          — isAnySkillPro: bool, proManualOverride: bool,
///                          anySkillProGrantedAt: Timestamp,
///                          avgResponseMinutes: int (optional, default = fast)
///
/// v15.x Phase 1 note: the `isAnySkillPro`, `proManualOverride`, and
/// `anySkillProGrantedAt` fields are blocked from owner writes by
/// firestore.rules. `checkAndRefreshProStatus` now delegates to a Cloud
/// Function (`evaluateMyProStatus` / `evaluateProStatusAsAdmin`) which
/// runs with the Admin SDK and bypasses the rule. The manual-override
/// admin actions (`setManualOverride`, `clearManualOverride`) keep
/// writing directly — they work because the admin's `|| isAdmin()`
/// rule clause grants full write access.
class ProService {
  ProService._();

  static final _db        = FirebaseFirestore.instance;
  static final _functions = FirebaseFunctions.instance;

  // ── Thresholds ────────────────────────────────────────────────────────────

  static Future<Map<String, dynamic>> fetchThresholds() async {
    final doc = await _db.collection('system_settings').doc('pro').get();
    final d = doc.data() ?? {};
    return {
      'minRating':           (d['minRating']           as num?)?.toDouble() ?? 4.8,
      'minOrders':           (d['minOrders']           as num?)?.toInt()    ?? 20,
      'maxResponseMinutes':  (d['maxResponseMinutes']  as num?)?.toInt()    ?? 15,
    };
  }

  static Stream<Map<String, dynamic>> streamThresholds() =>
      _db.collection('system_settings').doc('pro').snapshots().map((s) {
        final d = s.data() ?? {};
        return {
          'minRating':          (d['minRating']          as num?)?.toDouble() ?? 4.8,
          'minOrders':          (d['minOrders']          as num?)?.toInt()    ?? 20,
          'maxResponseMinutes': (d['maxResponseMinutes'] as num?)?.toInt()    ?? 15,
        };
      });

  static Future<void> saveThresholds({
    required double minRating,
    required int    minOrders,
    required int    maxResponseMinutes,
  }) =>
      _db.collection('system_settings').doc('pro').set({
        'minRating':          minRating,
        'minOrders':          minOrders,
        'maxResponseMinutes': maxResponseMinutes,
      }, SetOptions(merge: true));

  // ── Eligibility check ─────────────────────────────────────────────────────

  /// Delegates Pro evaluation to the server. The client no longer writes
  /// `isAnySkillPro` directly — firestore.rules block owner writes on
  /// that field. Instead:
  ///
  ///   * uid == current-user  → calls `evaluateMyProStatus` (no params).
  ///   * uid != current-user  → calls `evaluateProStatusAsAdmin({targetUid})`
  ///                            (server-checks admin role).
  ///
  /// Returns the evaluated badge state. Throws FirebaseFunctionsException
  /// on permission/rate-limit failures so the caller can surface a toast.
  /// Empty uid short-circuits to `false` like the legacy API.
  static Future<bool> checkAndRefreshProStatus(String uid) async {
    if (uid.isEmpty) return false;

    final currentUid = FirebaseAuth.instance.currentUser?.uid;
    final isSelf     = currentUid != null && currentUid == uid;

    final callable = _functions.httpsCallable(
      isSelf ? 'evaluateMyProStatus' : 'evaluateProStatusAsAdmin',
    );
    final res = await callable.call(isSelf ? null : {'targetUid': uid});
    final data = (res.data as Map?)?.cast<String, dynamic>() ?? const {};
    return data['isPro'] == true;
  }

  // ── Manual override ───────────────────────────────────────────────────────

  /// Admin grants or revokes Pro manually. Sets `proManualOverride: true` so
  /// the automatic check won't overwrite the decision.
  static Future<void> setManualOverride(String uid, {required bool isPro}) async {
    await _db.collection('users').doc(uid).update({
      'isAnySkillPro':      isPro,
      'proManualOverride':  true,
    });
    CachedReaders.invalidateProvider(uid); // §61
  }

  /// Removes the manual override so automatic checks resume.
  static Future<void> clearManualOverride(String uid) async {
    await _db.collection('users').doc(uid).update({
      'proManualOverride': false,
    });
    CachedReaders.invalidateProvider(uid); // §61
  }

  // ── Provider metrics snapshot ─────────────────────────────────────────────

  /// Returns a full [ProMetrics] snapshot for the given [uid].
  /// Used by the provider-facing AI insights screen.
  static Future<ProMetrics> fetchProviderMetrics(String uid) async {
    final results = await Future.wait([
      _db.collection('users').doc(uid).get(),
      fetchThresholds(),
      _countCompletedOrders(uid),
      _countRecentExpertCancellations(uid),
    ]);

    final userData      = (results[0] as DocumentSnapshot).data()
            as Map<String, dynamic>? ?? {};
    final thresholds    = results[1] as Map<String, dynamic>;
    final completedCnt  = results[2] as int;
    final cancelCnt     = results[3] as int;

    return ProMetrics(
      rating:                 (userData['rating']             as num?)?.toDouble() ?? 0.0,
      avgResponseMinutes:     (userData['avgResponseMinutes'] as num?)?.toInt()    ?? 0,
      completedOrders:        completedCnt,
      recentCancellations:    cancelCnt,
      isManualOverride:       userData['proManualOverride']  == true,
      isAnySkillPro:          userData['isAnySkillPro']      == true,
      anySkillProGrantedAt:   (userData['anySkillProGrantedAt'] as Timestamp?)?.toDate(),
      thresholdMinRating:     thresholds['minRating']        as double,
      thresholdMinOrders:     thresholds['minOrders']        as int,
      thresholdMaxResponseMins: thresholds['maxResponseMinutes'] as int,
    );
  }

  // ── Internal helpers ──────────────────────────────────────────────────────

  static Future<int> _countCompletedOrders(String uid) async {
    final snap = await _db
        .collection('jobs')
        .where('expertId', isEqualTo: uid)
        .where('status',   isEqualTo: 'completed')
        .get();
    return snap.docs.length;
  }

  /// Counts jobs where the expert cancelled within the last 30 days.
  /// Filters cancelledBy + cancelledAt client-side to avoid composite indexes.
  static Future<int> _countRecentExpertCancellations(String uid) async {
    final cutoff = DateTime.now().subtract(const Duration(days: 30));
    final snap = await _db
        .collection('jobs')
        .where('expertId', isEqualTo: uid)
        .where('status',   isEqualTo: 'cancelled')
        .get();
    int count = 0;
    for (final doc in snap.docs) {
      final d = doc.data();
      if (d['cancelledBy'] != 'expert') continue;
      final ts = d['cancelledAt'];
      DateTime? cancelledAt;
      if (ts is Timestamp) cancelledAt = ts.toDate();
      if (cancelledAt != null && cancelledAt.isAfter(cutoff)) count++;
    }
    return count;
  }
}

// ── Provider metrics snapshot ─────────────────────────────────────────────────

class ProMetrics {
  final double    rating;
  final int       avgResponseMinutes;
  final int       completedOrders;
  final int       recentCancellations;
  final bool      isManualOverride;
  final bool      isAnySkillPro;
  /// When the badge was granted (null if never held). Refreshed on every
  /// new grant — historical grants live in `admin_audit_log.previousGrantedAt`.
  final DateTime? anySkillProGrantedAt;
  final double    thresholdMinRating;
  final int       thresholdMinOrders;
  final int       thresholdMaxResponseMins;

  const ProMetrics({
    required this.rating,
    required this.avgResponseMinutes,
    required this.completedOrders,
    required this.recentCancellations,
    required this.isManualOverride,
    required this.isAnySkillPro,
    this.anySkillProGrantedAt,
    required this.thresholdMinRating,
    required this.thresholdMinOrders,
    required this.thresholdMaxResponseMins,
  });

  bool get ratingOk    => rating >= thresholdMinRating;
  bool get ordersOk    => completedOrders >= thresholdMinOrders;
  bool get responseOk  => avgResponseMinutes == 0 || avgResponseMinutes <= thresholdMaxResponseMins;
  bool get cancelOk    => recentCancellations == 0;
  bool get eligibleForPro => ratingOk && ordersOk && responseOk && cancelOk;

  /// 0.0–1.0 overall progress toward Pro (average of 4 criteria).
  double get overallProgress {
    double score = 0;
    score += ratingOk    ? 1.0 : (rating / thresholdMinRating).clamp(0.0, 1.0);
    score += ordersOk    ? 1.0 : (completedOrders / thresholdMinOrders).clamp(0.0, 1.0);
    score += responseOk  ? 1.0 : 0.3;  // partial credit for response time
    score += cancelOk    ? 1.0 : 0.0;
    return score / 4.0;
  }
}

File: functions/pro_service.js
// ═══════════════════════════════════════════════════════════════════════════
// AnySkill Pro — server-side evaluation helper (Phase 1, v15.x)
//
// Single source of truth for the badge-grant decision. Mirrors the Dart
// client logic in lib/services/pro_service.dart :: ProService
// .checkAndRefreshProStatus, but runs in Cloud Functions with the Admin
// SDK so it can write `isAnySkillPro` after the firestore.rules lock-down.
//
// Public surface: only `evaluateProStatus({db, uid, source, ...})`.
// All triggers + callables in functions/index.js call into here so the
// decision logic stays in one place.
//
// Audit log: every grant/revoke transition writes one document to
// `admin_audit_log/{auto-id}` so the admin dashboard + Watchtower can
// show "what happened, why, and who/what triggered it" without any
// extra plumbing.
// ═══════════════════════════════════════════════════════════════════════════

const admin = require("firebase-admin");
const { FieldValue } = require("firebase-admin/firestore");
const {
  buildGrantEmail,
  buildRevokeEmail,
  getRevocationCopy,
} = require("./pro_email_templates");

const ONE_DAY_MS    = 24 * 60 * 60 * 1000;
const THIRTY_D_MS   = 30 * ONE_DAY_MS;
const APP_LINK      = "https://anyskill-6fdf3.web.app";

// ── Default thresholds (match lib/services/pro_service.dart fallback) ──────
// Used when system_settings/pro doesn't exist or is missing keys.
const DEFAULT_THRESHOLDS = Object.freeze({
  minRating:           4.8,
  minOrders:           20,
  maxResponseMinutes:  15,
});

/**
 * Evaluate one provider's Pro eligibility and persist the result if it
 * changed. Idempotent — calling twice on a stable state is a no-op.
 *
 * @param {object}   args
 * @param {admin.firestore.Firestore} args.db
 * @param {string}   args.uid              Provider uid to evaluate.
 * @param {string}   args.source           One of: 'auto' | 'cron' | 'callable_self' | 'callable_admin'.
 * @param {string=}  args.triggerReason    Free-text context (e.g. 'job_completed:abc123').
 * @param {string=}  args.adminUid         Caller uid when source involves an admin.
 * @returns {Promise<{
 *   isPro: boolean,
 *   transition: 'granted'|'revoked'|'unchanged'|'manual_override_skip',
 *   skipped?: string,
 *   revokeReason?: string|null,
 * }>}
 */
async function evaluateProStatus({ db, uid, source = "auto", triggerReason = null, adminUid = null }) {
  if (!uid || typeof uid !== "string") {
    return { isPro: false, transition: "unchanged", skipped: "empty-uid" };
  }

  const [userSnap, thrSnap, completedCnt, cancelCnt] = await Promise.all([
    db.collection("users").doc(uid).get(),
    db.collection("system_settings").doc("pro").get(),
    _countCompletedOrders(db, uid),
    _countRecentExpertCancellations(db, uid),
  ]);

  if (!userSnap.exists) {
    return { isPro: false, transition: "unchanged", skipped: "no-user" };
  }
  const userData = userSnap.data() || {};

  // Manual override freezes the badge — admin decisions outrank auto-eval.
  if (userData.proManualOverride === true) {
    return {
      isPro: userData.isAnySkillPro === true,
      transition: "manual_override_skip",
    };
  }

  const thrData = thrSnap.exists ? (thrSnap.data() || {}) : {};
  const minRating       = _num(thrData.minRating,          DEFAULT_THRESHOLDS.minRating);
  const minOrders       = _num(thrData.minOrders,          DEFAULT_THRESHOLDS.minOrders);
  const maxResponseMins = _num(thrData.maxResponseMinutes, DEFAULT_THRESHOLDS.maxResponseMinutes);

  const rating          = _num(userData.rating,             0);
  const avgResponseMins = _num(userData.avgResponseMinutes, 0);

  // avgResponseMinutes == 0 → no data yet → don't penalise.
  const responseOk = avgResponseMins === 0 || avgResponseMins <= maxResponseMins;

  const isPro = rating >= minRating &&
    completedCnt >= minOrders &&
    responseOk &&
    cancelCnt === 0;

  const wasPro = userData.isAnySkillPro === true;
  if (wasPro === isPro) {
    return { isPro, transition: "unchanged" };
  }

  // ── State transition: write the user doc + audit-log entry ──────────────
  const previousGrantedAt = userData.anySkillProGrantedAt || null;
  const updatePayload = { isAnySkillPro: isPro };
  if (isPro) {
    // Per product decision (v15.x): refresh the timestamp on every new
    // grant. Historical grants live in admin_audit_log.previousGrantedAt.
    updatePayload.anySkillProGrantedAt = FieldValue.serverTimestamp();
  }
  // We deliberately do NOT clear anySkillProGrantedAt on revoke. The field
  // then represents "last held the badge" until the next grant.

  await db.collection("users").doc(uid).update(updatePayload);

  // Identify the criterion that broke (only meaningful on revoke).
  const revokeReason = isPro ? null : _identifyRevokeReason({
    rating, completedCnt, avgResponseMins, cancelCnt,
    minRating, minOrders, maxResponseMins,
  });

  await db.collection("admin_audit_log").add({
    action:           isPro ? "pro_granted" : "pro_revoked",
    targetUserId:     uid,
    targetUserName:   userData.name || userData.fullName || null,
    source,                  // 'auto' | 'cron' | 'callable_self' | 'callable_admin'
    triggerReason,           // optional context (e.g. 'job_completed:JOBID')
    adminUid,                // populated when an admin triggered the eval
    revocationReason: revokeReason,
    metricsSnapshot: {
      rating,
      completedOrders:     completedCnt,
      avgResponseMinutes:  avgResponseMins,
      recentCancellations: cancelCnt,
      thresholds: { minRating, minOrders, maxResponseMinutes: maxResponseMins },
    },
    previousGrantedAt,       // null on first grant; was-Date on re-grant
    createdAt:        FieldValue.serverTimestamp(),
    expireAt:         _ttl30Days(),  // §19 — admin_audit_log NOT TTL'd, but we
                                      // include expireAt as a forward-compat
                                      // hint. Removing the field is a one-line
                                      // change if policy reverses.
  });

  // ── Phase 2: notify the provider ──────────────────────────────────────
  // Three side-effects, all best-effort and isolated in try/catch so a
  // failure in one (e.g. missing FCM token, mail extension down) doesn't
  // abort the others or roll back the audit-log entry.
  const providerName = userData.name || userData.fullName || "שלום";
  const fcmToken     = userData.fcmToken || userData.deviceToken || null;
  const email        = userData.email || null;

  await _notifyProviderTransition({
    db,
    uid,
    providerName,
    fcmToken,
    email,
    isPro,
    metrics: {
      rating,
      completedDeals:   completedCnt,
      avgResponseTime:  avgResponseMins,
      cancellations:    cancelCnt,
      thresholds: { minRating, minOrders, maxResponseMinutes: maxResponseMins },
    },
  });

  return {
    isPro,
    transition: isPro ? "granted" : "revoked",
    revokeReason,
  };
}

// ── Notification fan-out (in-app + push + email) ───────────────────────────
// Best-effort — every channel wrapped in its own try/catch so a partial
// outage doesn't abort the whole transition. The audit log is the durable
// record; everything below is delivery.
async function _notifyProviderTransition({
  db, uid, providerName, fcmToken, email, isPro, metrics,
}) {
  const title = isPro
    ? "🏆 קיבלת את תג AnySkill Pro!"
    : "💙 עדכון לגבי תג AnySkill Pro שלך";
  const body = isPro
    ? `מזל טוב, ${providerName}! הצטרפת למועדון Pro של AnySkill.`
    : `${providerName}, התג הוסר זמנית. תוכל/י לחזור אליו — בדוק/י את הדשבורד לפרטים.`;

  // 1. In-app notification (always written — durable record).
  try {
    await db.collection("notifications").add({
      userId:    uid,
      title,
      body,
      type:      isPro ? "pro_granted" : "pro_revoked",
      isRead:    false,
      createdAt: FieldValue.serverTimestamp(),
    });
  } catch (e) {
    console.error(`[pro_notify] notifications doc failed for ${uid}:`, e.message);
  }

  // 2. FCM push (skipped gracefully if no token).
  if (fcmToken) {
    try {
      await admin.messaging().send({
        token: fcmToken,
        notification: { title, body },
        android: {
          priority: "high",
          notification: { channelId: "anyskill_default" },
        },
        apns: {
          headers: { "apns-priority": "10" },
          payload: { aps: { sound: "default", contentAvailable: 1 } },
        },
        webpush: {
          notification: { icon: "/icons/Icon-192.png" },
          fcm_options:  { link: APP_LINK },
        },
        data: {
          type: isPro ? "pro_granted" : "pro_revoked",
          uid,
        },
      });
    } catch (e) {
      console.error(`[pro_notify] push failed for ${uid}:`, e.message);
    }
  } else {
    console.warn(`[pro_notify] uid=${uid} has no FCM token, push skipped`);
  }

  // 3. Email via Firebase Trigger Email extension (mail/{auto-id}).
  if (email && typeof email === "string" && email.includes("@")) {
    try {
      const tpl = isPro
        ? buildGrantEmail({ providerName, appLink: APP_LINK })
        : (() => {
            const copy = getRevocationCopy(metrics);
            return buildRevokeEmail({
              providerName,
              revocationReason: copy.reason,
              currentRating:    metrics.rating,
              completedDeals:   metrics.completedDeals,
              avgResponseTime:  metrics.avgResponseTime,
              cancellations:    metrics.cancellations,
              recoveryTip:      copy.tip,
              appLink:          APP_LINK,
            });
          })();
      await db.collection("mail").add({
        to:      email,
        message: { subject: tpl.subject, html: tpl.html },
      });
    } catch (e) {
      console.error(`[pro_notify] mail enqueue failed for ${uid}:`, e.message);
    }
  } else {
    console.warn(`[pro_notify] uid=${uid} has no email, skipping email`);
  }
}

// ── Internal helpers ───────────────────────────────────────────────────────

function _num(v, fallback) {
  if (typeof v === "number" && Number.isFinite(v)) return v;
  if (typeof v === "string" && v.trim() !== "") {
    const n = Number(v);
    if (Number.isFinite(n)) return n;
  }
  return fallback;
}

function _ttl30Days() {
  return admin.firestore.Timestamp.fromDate(new Date(Date.now() + THIRTY_D_MS));
}

async function _countCompletedOrders(db, uid) {
  const snap = await db.collection("jobs")
    .where("expertId", "==", uid)
    .where("status",   "==", "completed")
    .get();
  return snap.size;
}

async function _countRecentExpertCancellations(db, uid) {
  const cutoff = Date.now() - THIRTY_D_MS;
  const snap = await db.collection("jobs")
    .where("expertId", "==", uid)
    .where("status",   "==", "cancelled")
    .get();
  let count = 0;
  for (const doc of snap.docs) {
    const d = doc.data();
    if (d.cancelledBy !== "expert") continue;
    const ts = d.cancelledAt;
    let cancelledAtMs = null;
    if (ts && typeof ts.toMillis === "function") {
      cancelledAtMs = ts.toMillis();
    }
    if (cancelledAtMs !== null && cancelledAtMs > cutoff) count++;
  }
  return count;
}

function _identifyRevokeReason({ rating, completedCnt, avgResponseMins, cancelCnt, minRating, minOrders, maxResponseMins }) {
  // Order matches user-spec recovery-tip priority: cancellations first
  // (they're the most actionable), then quantitative shortfalls.
  if (cancelCnt > 0) {
    return `expert_cancellation_30d (count=${cancelCnt})`;
  }
  if (rating < minRating) {
    return `rating_below_threshold (current=${rating.toFixed(2)}, min=${minRating})`;
  }
  if (completedCnt < minOrders) {
    return `insufficient_orders (current=${completedCnt}, min=${minOrders})`;
  }
  if (avgResponseMins > maxResponseMins) {
    return `slow_response (current=${avgResponseMins}min, max=${maxResponseMins}min)`;
  }
  return "unknown";
}

module.exports = { evaluateProStatus };
